Add conformance CI, update and fix tests

mvdbeek opened the pull request.

jmchilton committed:
Rewrite CWL tools.

mvdbeek committed:
Just pass along ephemeral_collection.persistent_object

mvdbeek committed:
Fallback to regular Galaxy code if not clearly in CWL mode

mvdbeek committed:
Fallback to regular Galaxy code if not clearly in CWL mode

mvdbeek committed:
Restore pulling out dataset instance for simple data inputs

mvdbeek committed:
Child collection can be None

mvdbeek committed:
Read secondary files from outputs/ directory

mvdbeek committed:
Print converted documents

mvdbeek committed:
Write out just a single marked test file

mvdbeek committed:
Reduce tests to single file per conformance suite

mvdbeek committed:
List all marks that CWL conformance tests use

jmchilton commented on lib/galaxy/job_execution/output_collect.py:
```diff
@@ -593,15 +593,16 @@ def default_exit_code_file(files_dir, id_tag):
 def collect_extra_files(object_store, dataset, job_working_directory):
     file_name = dataset.dataset.extra_files_path_name_from(object_store)
-    temp_file_path = os.path.join(job_working_directory, "working", file_name)
+    output_dir = "working" if os.path.exists(os.path.join(job_working_directory, "working", file_name)) else "outputs"
```
I was going to augment tools with information about how to annotate this 😆 - a whole day project. This is so much easier, really awesome!

mvdbeek replied:
We should ultimately switch everything, and then keep that logic in for a release or two :).

mvdbeek committed:
Add top level cwl_conformance marker

mvdbeek committed:
Add run_tests.sh -cwl

mvdbeek committed:
Include version in conformance marker

mvdbeek committed:
Add CWL conformance testing via GH workflows

nsoranzo committed:
Update conformance tests from upstream CWL repos

Also:
- Update paths
- Removed unused files.

mvdbeek committed:
Fix run_tests.sh for tests without marks

mvdbeek committed:
Lint run_cwl_conformance.oy

mvdbeek committed:
Lint/remove dead code in run_cwl.py

mvdbeek committed:
Various linting fixes

mvdbeek committed:
Drop API tests that are also covered as conformance tests

mvdbeek commented on run_tests.sh:
```diff
@@ -675,10 +675,11 @@ if [ "$test_script" = 'pytest' ]; then
-        "$test_script" -v --html "$report_file" $coverage_arg  $xunit_args $extra_args "$@"
+        marker_arg=""
     fi
+    "$test_script" -v --html "$report_file" $coverage_arg $xunit_args $extra_args $marker_arg "$@"
```
I think I've spent an hour yesterday trying everything I could think of to get `"not cwl_conformance"` as an argument. If you don't quote this you get ` -m ''\''not' 'cwl_conformance'\'''`, if you do you get `'-m '\''not cwl_conformance'\'''`.

mvdbeek committed:
Only upload html report on test failure

mvdbeek committed:
Exclude CWL tools and auto-generated conformance tests from linting

nsoranzo committed:
Remove broken tox testenvs

nsoranzo committed:
Create separate artifacts for CWL CI matrix elements

nsoranzo committed:
Create self-contained pytest reports

Also:
- refactor handling of markers, need to use eval unfortunately
- fix issues reported by `shellcheck`

nsoranzo committed:
Ignore run_cwl_tests.html

nsoranzo committed:
Update CWL 1.0 conformance tests

Also add `update-cwl-conformance-tests` a Makefile target.

mvdbeek committed:
Update new green tests

jmchilton commented on test/unit/tools/cwl_tools/conformance_to_test_cases.py:
```diff
@@ -163,6 +192,7 @@ class CwlConformanceTestCase(BaseCwlWorklfowTestCase):
         "valuefrom_ignored_null",
     ],
     "v1.1": [
+        "v1_1_wf_simple",
```
Wrong syntax here.

nsoranzo committed:
Update CWL 1.2 conformance tests

jmchilton commented:
Amazing - thanks to you both for this awesome work!
